File: experiments/adrianco/experiment-15-sonnet5/brazil/runs/language=typescript_model=opus-4.8_prompt=bdd/rep1/src/format.ts
/**
 * Human-readable formatting of query results, mirroring the answer formats in
 * the specification. Each MCP tool returns one of these strings as its text
 * content so an LLM (or a person) sees a clean, self-describing answer.
 */
import { formatDate } from "./normalize.js";
import type { Match, Player, StandingsRow } from "./types.js";
import type { HeadToHead } from "./queries/matches.js";
import type { TeamStats } from "./queries/teams.js";
import type { MatchAggregate, BiggestWin, TeamGoalTotal } from "./queries/stats.js";
import type { ClubSummary } from "./queries/players.js";

/** Format a competition/round tag like "(Brasileirão Série A Round 22)". */
function contextTag(match: Match): string {
  const parts = [match.competition];
  if (match.round) parts.push(`Round ${match.round}`);
  else if (match.stage) parts.push(match.stage);
  return `(${parts.join(" ")})`;
}

/** One-line scoreline, e.g. "2023-09-03: Flamengo 2-1 Fluminense (Brasileirão Round 22)". */
export function formatMatchLine(match: Match): string {
  const score =
    match.homeGoals === null || match.awayGoals === null
      ? "vs"
      : `${match.homeGoals}-${match.awayGoals}`;
  return `${formatDate(match.date)}: ${match.homeTeam} ${score} ${match.awayTeam} ${contextTag(match)}`;
}

export function formatMatchList(matches: Match[], heading: string, limit = 25): string {
  if (matches.length === 0) return `${heading}\nNo matches found.`;
  const shown = matches.slice(0, limit).map((m) => `- ${formatMatchLine(m)}`);
  const lines = [heading, ...shown];
  if (matches.length > limit) {
    lines.push(`- ... (${matches.length - limit} more matches in dataset)`);
  }
  return lines.join("\n");
}

export function formatHeadToHead(record: HeadToHead, matches: Match[], limit = 10): string {
  const heading = `${record.teamA} vs ${record.teamB} head-to-head:`;
  const recent = matches.slice(0, limit).map((m) => `- ${formatMatchLine(m)}`);
  const summary =
    `\nHead-to-head in dataset (${record.matches} matches): ` +
    `${record.teamA} ${record.teamAWins} wins, ${record.teamB} ${record.teamBWins} wins, ${record.draws} draws.` +
    `\nGoals: ${record.teamA} ${record.teamAGoals}, ${record.teamB} ${record.teamBGoals}.`;
  if (matches.length === 0) {
    return `${heading}\nNo matches found between these teams in the dataset.`;
  }
  const extra = matches.length > limit ? `\n- ... (${matches.length - limit} more)` : "";
  return `${heading}\n${recent.join("\n")}${extra}${summary}`;
}

function pct(rate: number): string {
  return `${(rate * 100).toFixed(1)}%`;
}

export function formatTeamStats(stats: TeamStats, scopeLabel: string): string {
  return [
    `${stats.team} record (${scopeLabel}):`,
    `- Matches: ${stats.played}`,
    `- Wins: ${stats.wins}, Draws: ${stats.draws}, Losses: ${stats.losses}`,
    `- Goals For: ${stats.goalsFor}, Goals Against: ${stats.goalsAgainst} (GD ${signed(stats.goalDifference)})`,
    `- Points: ${stats.points}`,
    `- Win rate: ${pct(stats.winRate)}`,
    `- Home: ${stats.home.wins}W ${stats.home.draws}D ${stats.home.losses}L | ` +
      `Away: ${stats.away.wins}W ${stats.away.draws}D ${stats.away.losses}L`,
    `- Competitions: ${stats.competitions.join(", ") || "none"}`,
  ].join("\n");
}

function signed(n: number): string {
  return n > 0 ? `+${n}` : `${n}`;
}

export function formatStandings(rows: StandingsRow[], heading: string, limit = 20): string {
  if (rows.length === 0) return `${heading}\nNo matches available to build the table.`;
  const lines = rows.slice(0, limit).map((r) => {
    const tag = r.position === 1 ? " - Champion" : "";
    return `${r.position}. ${r.team} - ${r.points} pts (${r.wins}W, ${r.draws}D, ${r.losses}L, GD ${signed(r.goalDifference)})${tag}`;
  });
  return [heading, ...lines].join("\n");
}

export function formatPlayer(player: Player): string {
  const rating = player.overall !== null ? `Overall: ${player.overall}` : "Overall: n/a";
  const pos = player.position || "?";
  const club = player.club || "Free agent";
  return `${player.name} - ${rating}, Position: ${pos}, Club: ${club}, Nationality: ${player.nationality}`;
}

export function formatPlayerList(players: Player[], heading: string, limit = 25): string {
  if (players.length === 0) return `${heading}\nNo players found.`;
  const lines = players.slice(0, limit).map((p, i) => `${i + 1}. ${formatPlayer(p)}`);
  if (players.length > limit) {
    lines.push(`... (${players.length - limit} more players)`);
  }
  return [heading, ...lines].join("\n");
}

export function formatClubSummaries(summaries: ClubSummary[], heading: string, limit = 15): string {
  if (summaries.length === 0) return `${heading}\nNo clubs found.`;
  const lines = summaries
    .slice(0, limit)
    .map(
      (s) =>
        `- ${s.club}: ${s.playerCount} players (avg rating: ${s.averageOverall}, top: ${s.topPlayer})`,
    );
  return [heading, ...lines].join("\n");
}

export function formatAggregate(agg: MatchAggregate, scopeLabel: string): string {
  return [
    `Match statistics (${scopeLabel}):`,
    `- Matches: ${agg.matches}`,
    `- Total goals: ${agg.totalGoals}`,
    `- Average goals per match: ${agg.averageGoalsPerMatch}`,
    `- Home wins: ${agg.homeWins} (${pct(agg.homeWinRate)})`,
    `- Away wins: ${agg.awayWins} (${pct(agg.awayWinRate)})`,
    `- Draws: ${agg.draws} (${pct(agg.drawRate)})`,
  ].join("\n");
}

export function formatBiggestWins(wins: BiggestWin[], heading: string): string {
  if (wins.length === 0) return `${heading}\nNo matches found.`;
  const lines = wins.map((w, i) => `${i + 1}. ${formatMatchLine(w.match)} [margin ${w.margin}]`);
  return [heading, ...lines].join("\n");
}

export function formatTopScoringTeams(totals: TeamGoalTotal[], heading: string): string {
  if (totals.length === 0) return `${heading}\nNo matches found.`;
  const lines = totals.map(
    (t, i) => `${i + 1}. ${t.team} - ${t.goalsFor} goals in ${t.matches} matches`,
  );
  return [heading, ...lines].join("\n");
}

File: experiments/adrianco/experiment-7/brazil/runs/language=clojure_model=claude-opus-4-8-fast/rep2/src/brazilian_soccer/format.clj
(ns brazilian-soccer.format
  "=============================================================================
   format.clj — Human-readable rendering of query results
   -----------------------------------------------------------------------------
   Context:
     The MCP tools return text content to the calling LLM/user. This namespace
     turns the plain-data results from brazilian-soccer.queries into the
     readable, example-style strings shown in the specification (match lines,
     standings tables, team records, player rankings, etc.).
   ============================================================================="
  (:require [clojure.string :as str]))

(defn- pct [x] (format "%.1f%%" (* 100.0 (or x 0.0))))

(defn match-line
  "Render a single match like:
     2023-09-03: Flamengo 2-1 Fluminense (Brasileirão, Round 22)"
  [{:keys [date home-team away-team home-goal away-goal competition round stage]}]
  (let [score (if (and home-goal away-goal)
                (str home-goal "-" away-goal)
                "?-?")
        extra (cond
                round (str ", Round " round)
                stage (str ", " stage)
                :else "")]
    (format "%s: %s %s %s (%s%s)"
            (or date "????-??-??") home-team score away-team competition extra)))

(defn matches-report
  "Render a list of matches with a header and optional truncation note."
  [matches {:keys [title shown]}]
  (if (empty? matches)
    (str (or title "Matches") ": none found.")
    (let [shown   (or shown (count matches))
          listed  (take shown matches)
          more    (- (count matches) shown)]
      (str (when title (str title "\n"))
           (str/join "\n" (map #(str "- " (match-line %)) listed))
           (when (pos? more) (format "\n- ... (%d more matches)" more))))))

(defn team-stats-report
  [{:keys [team season competition venue played wins draws losses
           goals-for goals-against win-rate]}]
  (let [scope (str/join " " (remove str/blank?
                                    [(when competition (str competition))
                                     (when season (str season))
                                     (when (not= venue :all) (str (clojure.core/name venue) " only"))]))]
    (str team " record" (when (seq scope) (str " (" scope ")")) ":\n"
         (format "- Matches: %d\n- Wins: %d, Draws: %d, Losses: %d\n- Goals For: %d, Goals Against: %d\n- Win rate: %s"
                 played wins draws losses goals-for goals-against (pct win-rate)))))

(defn head-to-head-report
  [{:keys [team1 team2 team1-wins team2-wins draws total matches]}]
  (str (format "%s vs %s — head-to-head (%d matches in dataset):\n" team1 team2 total)
       (format "%s %d wins, %s %d wins, %d draws" team1 team1-wins team2 team2-wins draws)
       (when (seq matches)
         (str "\n\nRecent meetings:\n"
              (str/join "\n" (map #(str "- " (match-line %)) (take 10 matches)))))))

(defn standings-report
  [rows {:keys [title limit] :or {limit 20}}]
  (if (empty? rows)
    (str (or title "Standings") ": no data.")
    (str (when title (str title "\n"))
         (str/join "\n"
                   (map (fn [{:keys [rank team points played wins draws losses goal-diff]}]
                          (format "%2d. %s - %d pts (%dW %dD %dL, GD %+d)"
                                  rank team points wins draws losses goal-diff))
                        (take limit rows))))))

(defn competition-stats-report
  [{:keys [competition season matches total-goals avg-goals
           home-wins away-wins draws home-win-rate]}]
  (str (str/join " " (remove nil? [competition season "statistics:"])) "\n"
       (format "- Matches: %d\n- Total goals: %d\n- Average goals per match: %.2f\n- Home wins: %d, Away wins: %d, Draws: %d\n- Home win rate: %s"
               matches total-goals avg-goals home-wins away-wins draws (pct home-win-rate))))

(defn players-report
  [players {:keys [title limit] :or {limit 25}}]
  (if (empty? players)
    (str (or title "Players") ": none found.")
    (str (when title (str title "\n"))
         (str/join "\n"
                   (map-indexed
                    (fn [i {:keys [name overall position club nationality age]}]
                      (format "%d. %s - Overall: %s, Position: %s, Club: %s, Nationality: %s%s"
                              (inc i) name (or overall "?") (or position "?")
                              (or club "Free agent") (or nationality "?")
                              (if age (str ", Age: " age) "")))
                    (take limit players))))))

(defn club-summary-report
  [rows {:keys [title limit] :or {limit 20}}]
  (if (empty? rows)
    (str (or title "Clubs") ": none found.")
    (str (when title (str title "\n"))
         (str/join "\n"
                   (map (fn [{:keys [club count avg-overall]}]
                          (format "- %s: %d players (avg rating: %.1f)" club count avg-overall))
                        (take limit rows))))))
